src/analysis/processing/qgsalgorithmflattenrelationships.cpp

```diff
@@ -0,0 +1,159 @@
+
+#include "qgsalgorithmflattenrelationships.h"
+#include "qgsrelationmanager.h"
+#include "qgsvectorlayer.h"
+#include "qgsvectorlayerfeatureiterator.h"
+
+///@cond PRIVATE
+
+QString QgsFlattenRelationshipsAlgorithm::name() const
+{
+  return QStringLiteral( "flattenrelationships" );
+}
+
+QString QgsFlattenRelationshipsAlgorithm::displayName() const
+{
+  return QObject::tr( "Flatten relationships" );
+}
+
+QStringList QgsFlattenRelationshipsAlgorithm::tags() const
+{
+  return QObject::tr( "join,export,single,table" ).split( ',' );
+}
+
+QString QgsFlattenRelationshipsAlgorithm::group() const
+{
+  return QObject::tr( "Vector general" );
+}
+
+QString QgsFlattenRelationshipsAlgorithm::groupId() const
+{
+  return QStringLiteral( "vectorgeneral" );
+}
+
+QString QgsFlattenRelationshipsAlgorithm::shortDescription() const
+{
+  return QObject::tr( "Flattens all relationships for a vector layer." );
+}
+
+QString QgsFlattenRelationshipsAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "This algorithm flattens all relationships for a vector layer, exporting a single layer "
+                      "containing one master feature per related feature. This master feature contains all "
+                      "the attributes for the related features." );
+}
+
+QgsProcessingAlgorithm::Flags QgsFlattenRelationshipsAlgorithm::flags() const
+{
+  return QgsProcessingAlgorithm::flags() | FlagRequiresProject;
+}
+
+void QgsFlattenRelationshipsAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterVectorLayer( QStringLiteral( "INPUT" ),
+                QObject::tr( "Input layer" ), QList< int>() << QgsProcessing::TypeVector ) );
+
+  addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "OUTPUT" ), QObject::tr( "Flattened layer" ), QgsProcessing::TypeVectorAnyGeometry, QVariant(), true, true ) );
+}
+
+QgsFlattenRelationshipsAlgorithm *QgsFlattenRelationshipsAlgorithm::createInstance() const
+{
+  return new QgsFlattenRelationshipsAlgorithm();
+}
+
+bool QgsFlattenRelationshipsAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback * )
+{
+  QgsProject *project = context.project();
+  if ( !project )
+    throw QgsProcessingException( QObject::tr( "No project available for relationships" ) );
+
+  QgsVectorLayer *layer = parameterAsVectorLayer( parameters, QStringLiteral( "INPUT" ), context );
+  if ( !layer )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "INPUT" ) ) );
+
+  const QList<QgsRelation> relations = project->relationManager()->referencedRelations( layer );
+  if ( relations.size() > 1 )
+    throw QgsProcessingException( QObject::tr( "Found %1 relations. This algorithm currently supports only a single relation." ).arg( relations.size() ) );
+
+  if ( relations.empty() )
+    throw QgsProcessingException( QObject::tr( "No relations found." ) );
+
+  mRelation = relations.at( 0 );
```

Comment:
While reading the description, I was expecting to get all the relations for this layer.
If this is not the case, this should be explicitly stated that the alg randomly takes the first relation in the list.

Reply:
Thanks, I've reworded the name/description to clarify this. I couldn't think of a clean way to handle more than one relationship for a layer -- exporting a full cross join between all relations seems very over the top, and unlikely what a user is wanting...